pymc_marketing/mmm/data_generator.py

```diff
@@ -0,0 +1,521 @@
+        df = self.df
+        self.is_lagged = True # signals that adstock has been applied
+        self.lag_fn = fn
+        self.alphas = np.random.beta(.5, .5, self.num_channels)
+        self.thetas = None
+        if fn == 'delayed':
+            # trying to keep delays between 0 and 3 for the most part
+            self.thetas = np.random.binomial(max_lag-1, .1, self.num_channels)
+        self.max_lag = max_lag
+
+        # Apply adstock transformation
+        for c in range(self.num_channels):
+            args = {
+                  'x':df[f"x{c}"].to_numpy()
+                , 'alpha':self.alphas[c]
+                , 'l_max':max_lag
+                , 'normalize':True
+                }
+            if fn == 'geometric':
+                fn = geometric_adstock
+            else:
+                fn = delayed_adstock
+                args['theta']=self.thetas[c]
+            df[f"x{c}_adstock"] = fn(**args).eval().flatten()
+
+
+    def plot_adstock_effect(self):
+        '''Plots the raw channels compared to when their adstock effects are applied.'''
+        assert self.is_lagged
+        fig, ax = self.plot_channel_spends(show=False)
+        self.plot_channel_spends(
+                suffix="_adstock", fig=fig, ax=ax, title='Adstock Effect', alpha=.5)
+
+
+    def add_saturation_effect(self, lambdas_scale:float=3.5, lambas_sd:float=1):
+        ''' Apply a logistic saturation transformation
+        Parameters
+        ----------
+        lambdas_scale: float
+            Mean of the normal distribution the lambdas are sampled from.
+        lambdas_sd: float
+            Standard deviation of the normal distribution the lambdas are sampled from.
+        ''';
+        self.is_saturated = True
+        # column reference will be "x1_adstock" if adstock has been applied
+        # or just "x1" if adstock was not opted for
+        lambdas = np.random.normal(lambdas_scale, lambas_sd, self.num_channels)
+        self.lambdas = lambdas
+
+        for i, λ in enumerate(lambdas):
+            col_ref = [f"x{i}",f"x{i}_adstock"][self.is_lagged]
+            col_name = col_ref+"_saturated"
+            self.df[col_name] = logistic_saturation(x=self.df[col_ref].to_numpy(), lam=λ).eval()
+
+
+    def plot_saturation_effect(self):
+        '''Plots the raw channels compared to saturation effects are applied.
+        If lagging has been applied, then the lagged signal is plotted in lieu of the raw signal.'''
+        assert self.is_saturated
+        if not self.is_lagged:
+            fig, ax = self.plot_channel_spends(show=False)
+            self.plot_channel_spends(
+                    suffix="_saturated", title="Saturation Effect", fig=fig, ax=ax, alpha=.5)
+        else:
+            fig, ax = self.plot_channel_spends(suffix="_adstock", show=False)
+            self.plot_channel_spends(
+                    suffix="_adstock_saturated", alpha=.5, title="Saturation Effect",
+                    fig=fig, ax=ax)
```

Comment:
These can be consolidate and by using if block to define different kwargs

Reply:
good call